cpp/src/parquet/statistics.cc

```diff
@@ -32,84 +32,362 @@ using arrow::MemoryPool;
+ public:
+  typedef typename DType::c_type T;
+
+  explicit TypedComparatorImpl(int type_length = -1) : type_length_(type_length) {}
+
+  bool CompareInline(const T& a, const T& b) {
+    return CompareHelper<DType, is_signed>::Compare(type_length_, a, b);
+  }
+
+  bool Compare(const T& a, const T& b) override { return CompareInline(a, b); }
+
+  void GetMinMax(const T* values, int64_t length, T* out_min, T* out_max) override {
+    T min = values[0];
+    T max = values[0];
+    for (int64_t i = 1; i < length; i++) {
+      if (CompareInline(values[i], min)) {
+        min = values[i];
+      } else if (CompareInline(max, values[i])) {
+        max = values[i];
+      }
+    }
+    *out_min = min;
+    *out_max = max;
+  }
+
+  void GetMinMaxSpaced(const T* values, int64_t length, const uint8_t* valid_bits,
+                       int64_t valid_bits_offset, T* out_min, T* out_max) override {
+    ::arrow::internal::BitmapReader valid_bits_reader(valid_bits, valid_bits_offset,
+                                                      length);
+    T min = values[0];
+    T max = values[0];
+    for (int64_t i = 0; i < length; i++) {
+      if (valid_bits_reader.IsSet()) {
+        if (CompareInline(values[i], min)) {
+          min = values[i];
+        } else if (CompareInline(max, values[i])) {
+          max = values[i];
+        }
+      }
+      valid_bits_reader.Next();
+    }
+    *out_min = min;
+    *out_max = max;
+  }
+
+ private:
+  int type_length_;
+};
+
+std::shared_ptr<Comparator> Comparator::Make(Type::type physical_type,
+                                             SortOrder::type sort_order,
+                                             int type_length) {
+  if (SortOrder::SIGNED == sort_order) {
+    switch (physical_type) {
+      case Type::BOOLEAN:
+        return std::make_shared<TypedComparatorImpl<BooleanType>>();
+      case Type::INT32:
+        return std::make_shared<TypedComparatorImpl<Int32Type>>();
+      case Type::INT64:
+        return std::make_shared<TypedComparatorImpl<Int64Type>>();
+      case Type::INT96:
+        return std::make_shared<TypedComparatorImpl<Int96Type>>();
+      case Type::FLOAT:
+        return std::make_shared<TypedComparatorImpl<FloatType>>();
+      case Type::DOUBLE:
+        return std::make_shared<TypedComparatorImpl<DoubleType>>();
+      case Type::BYTE_ARRAY:
+        return std::make_shared<TypedComparatorImpl<ByteArrayType>>();
+      case Type::FIXED_LEN_BYTE_ARRAY:
+        return std::make_shared<TypedComparatorImpl<FLBAType>>(type_length);
+      default:
+        ParquetException::NYI("Signed Compare not implemented");
+    }
+  } else if (SortOrder::UNSIGNED == sort_order) {
+    switch (physical_type) {
+      case Type::INT32:
+        return std::make_shared<TypedComparatorImpl<Int32Type, false>>();
+      case Type::INT64:
+        return std::make_shared<TypedComparatorImpl<Int64Type, false>>();
+      case Type::INT96:
```

Comment:
Remove?

Reply:
I'm leaving this so it's possible to instantiate the Comparator